python/ql/lib/semmle/python/frameworks/Gradio.qll

```diff
@@ -0,0 +1,123 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `gradio` PyPI package.
+ * See https://pypi.org/project/gradio/.
+ */
+
+import python
+import semmle.python.dataflow.new.RemoteFlowSources
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.ApiGraphs
+
+/**
+ * Provides models for the `gradio` PyPI package.
+ * See https://pypi.org/project/gradio/.
+ */
+module Gradio {
+  /**
+   * The event handlers, Interface and gradio.ChatInterface classes, which take untrusted data.
+   */
+  class GradioInput extends API::CallNode {
+    GradioInput() {
+      this =
+        API::moduleImport("gradio")
+            .getMember([
+                "Button", "Textbox", "UploadButton", "Slider", "JSON", "HTML", "Markdown", "File",
+                "AnnotatedImage", "Audio", "BarPlot", "Chatbot", "Checkbox", "CheckboxGroup",
+                "ClearButton", "Code", "ColorPicker", "Dataframe", "Dataset", "DownloadButton",
+                "Dropdown", "DuplicateButton", "FileExplorer", "Gallery", "HighlightedText",
+                "Image", "ImageEditor", "Label", "LinePlot", "LoginButton", "LogoutButton",
+                "Model3D", "Number", "ParamViewer", "Plot", "Radio", "ScatterPlot", "SimpleImage",
```

Comment:
Are all these vulnerable? I quickly checked a few of them and it looks like:
- `HighlightedText` only allows the user to specify substrings of a given text
- `ColorPicker` only passes on hex strings
- `AnnotatedImage` does not accept user input at all

We should probably only include actually vulnerable inputs here in order to avoid the query getting noisy.

Reply:
It’s the event listeners that take untrusted user input, so, for example, `gradio.AnnotatedImage.select(fn, inputs, outputs)` (see [docs](https://www.gradio.app/docs/gradio/annotatedimage#event-listeners)) and that's what we [model](https://github.com/github/codeql/pull/16135/files#diff-a9e4a77625b0915aef61917705ea26225a5aefd382cedf40ca9dedc69df782acR23-R38) here.

To clarify - with the above models, we look for any [event listeners](https://github.com/github/codeql/pull/16135/files#diff-a9e4a77625b0915aef61917705ea26225a5aefd382cedf40ca9dedc69df782acR33-R37) of the listed [classes](https://github.com/github/codeql/pull/16135/files#diff-a9e4a77625b0915aef61917705ea26225a5aefd382cedf40ca9dedc69df782acR24-R30). For example, `gr.Button` has only one, `click()` (see this [example](https://www.gradio.app/docs/gradio/blocks#example-usage), which asks for a user's name and displays it). Side note that if an event listener does not exist for a given class, CodeQL doesn't (can't) match it  (for example there is no `gr.Button.select()`). I found this way of modeling to be the most succinct and it doesn't slow performance, but I'm open to feedback around it.

Reply:
I think it is fine to model situations that will never match, I was more worried about false positives. But on reflection, I think this is fine to start; we can refine it if we actually encounter any problems..

Reply:
I've already found a few vulnerabilities using these models, including in stable-diffusion-webui (120k stars):
https://securitylab.github.com/advisories/GHSL-2024-010_stable-diffusion-webui/
https://securitylab.github.com/advisories/GHSL-2024-019_GHSL-2024-024_kohya_ss/
To be fair, I tested the `AnnotatedImage.select()` event listener (and several other event listeners on the list), and it does take user input, but it's not very popular to use, so I haven't actually seen vulnerabilities around it. Most of the vulnerabilities happen with `Button.click()` (including the 7 vulns I linked to above). I think there shouldn't be many false positives here, but I'll be happy to help if anything needs refining.